domain/schema/controller.go

```diff
@@ -591,3 +592,83 @@ CREATE TABLE flag (
 );
 `)
 }
+
+func userPermissionSchema() schema.Patch {
+	return schema.MakePatch(`
+CREATE TABLE permission_access_type (
+    id     INT PRIMARY KEY,
+    type   TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_permission_access_type
+ON permission_access_type (type);
+
+-- Maps to the Access type in core/permission package.
+INSERT INTO permission_access_type VALUES
+    (0, 'read'),
+    (1, 'write'),
+    (2, 'consume'),
+    (3, 'admin'),
+    (4, 'login'),
+    (5, 'addmodel'),
+    (6, 'superuser');
+
+CREATE TABLE permission_object_type (
+    id    INT PRIMARY KEY,
+    type  TEXT NOT NULL
+);
+
+CREATE UNIQUE INDEX idx_permission_object_type
+ON permission_object_type (type);
+
+INSERT INTO permission_object_type VALUES
+    (0, 'controller'),
+    (1, 'cloud'),
+    (2, 'model'),
+    (3, 'offer');
+
+CREATE TABLE permission_object_access (
+    id              INT PRIMARY KEY,
+    access_type     INT NOT NULL,
+    object_type     INT NOT NULL,
+    CONSTRAINT      uc_object_access UNIQUE (access_type, object_type),
+    CONSTRAINT      fk_permission_access_type
+        FOREIGN KEY (access_type)
+        REFERENCES  permission_access_type(id),
+    CONSTRAINT      fk_permission_object_type
+        FOREIGN KEY (object_type)
+        REFERENCES  permission_object_type(id)
+);
+
+CREATE UNIQUE INDEX idx_permission_object_access
+ON permission_object_access (access_type, object_type);
+
+INSERT INTO permission_object_access VALUES
+    (0, 5 ,0), -- login, controller
+    (1, 7 ,0), -- superuser, controller
+    (2, 4, 1), -- admin, cloud
+    (3, 6, 1), -- addmodel, cloud
+    (4, 1, 2), -- read, model
+    (5, 2, 2), -- write, model
+    (6, 4, 2), -- admin, model
+    (7, 1, 3), -- read, offer
+    (8, 3, 3), -- consume, offer
+    (9, 4, 3); -- admin, offer
+
+
+CREATE TABLE user_permission (
+    uuid            TEXT PRIMARY KEY,
+    object_name     TEXT NOT NULL, -- name or uuid of the object
+    object_access   INT NOT NULL,
+    user_uuid       TEXT NOT NULL,
+    -- ensure that for a user, only one object_name enters the DB
+    CONSTRAINT      uc_permission UNIQUE (user_uuid, object_name),
```

Comment:
Convention to-date is to create indexes instead of constraints for uniqueness. This will create a unique index under the hood.

By explicitly creating the index we can enforce _our_ naming convention on it.

As for this restriction, there's an implication that the verbs are sub/supersets of each other (write implies read for example). I'm not sure this is a blanket truth, but we'll find out soon enough.

I'm also not 100% sure of using the `object_access_id` over the `access_type_id`. If we have an enumeration in the code for the access types (as we do) we can skip a join when querying permissions.

Reply:
> there's an implication that the verbs are sub/supersets of each other (write implies read for example)

It's baked into the core package: https://github.com/juju/juju/blob/62efec59b2402498bdc2f28b77d3dc8aeabadcc4/core/permission/access.go#L135